components/places/src/storage/bookmarks.rs

```diff
@@ -1000,6 +1272,53 @@ fn get_raw_bookmark(db: &impl ConnExt, guid: &SyncGuid) -> Result<Option<RawBook
 mod tests {
     use super::*;
     use crate::db::PlacesDb;
+    use serde_json::Value;
+
+    fn insert_json_tree(conn: &PlacesDb, jtree: Value) {
+        let tree: BookmarkTreeNode = serde_json::from_value(jtree).expect("should be valid");
+        let folder_node = match tree {
+            BookmarkTreeNode::Folder(folder_node) => folder_node,
+            _ => panic!("must be a folder"),
+        };
+        insert_tree(conn, &folder_node).expect("should insert");
+    }
+
+    fn assert_json_tree(conn: &PlacesDb, folder: &SyncGuid, expected: Value) {
+        let fetched = fetch_tree(conn, folder).expect("should work").unwrap();
+        let deser_tree: BookmarkTreeNode = serde_json::from_value(expected).unwrap();
+        assert_eq!(fetched, deser_tree);
+    }
+
+    fn get_pos(conn: &PlacesDb, guid: &SyncGuid) -> u32 {
+        get_raw_bookmark(conn, guid)
+            .expect("should work")
+            .unwrap()
+            .position
+    }
+
+    // check the positions for children in a folder are "correct" in that
+    // the first child has a value of zero, etc - ie, this will fail if there
+    // are holes or duplicates in the position values.
```

Comment:
This function looks good as it is, but, if you're interested, Places has a [neat way](https://searchfox.org/mozilla-central/rev/92d11a33250a8e368c8ca3e962e15ca67117f765/toolkit/components/places/tests/bookmarks/test_bookmarks_reorder.js#114-121) to detect holes and dupes. 😄

Reply:
oh, that is nice! Done.